firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,762 @@
+            layer_extents[p, 3] = min(layer_extents[p, 3], cell_extents[cell, 1])
+    CHKERR(DMPlexRestoreTransitiveClosure(dm.dm, 0, PETSC_TRUE, NULL, &closure))
+    if mesh.comm.size == 1:
+        return layer_extents
+
+    # OK, now we have the correct extents for owned points, but
+    # potentially incorrect extents for ghost points, so do a SF Bcast
+    # over the point SF to get it right.
+    # TODO: this only works because we have a cell halo that contains
+    # all the cells that touch the vertices we own.  If this changes,
+    # e.g we don't grow the halo overlap, then we need to fix it.
+    out_extents = numpy.copy(layer_extents)
+    try:
+        tdict = MPI.__TypeDict__
+    except AttributeError:
+        tdict = MPI._typedict
+    typ = tdict[out_extents.dtype.char]
+    contig = typ.Create_contiguous(4)
+    contig.Commit()
+    sf = dm.getPointSF()
+    CHKERR(PetscSFBcastBegin(sf.sf, contig.ob_mpi,
+                             <const void*>layer_extents.data,
+                             <void *>out_extents.data))
+    CHKERR(PetscSFBcastEnd(sf.sf, contig.ob_mpi,
+                             <const void*>layer_extents.data,
+                             <void *>out_extents.data))
+    contig.Free()
+    return out_extents
+
+
+@cython.wraparound(False)
+def create_section(mesh, nodes_per_entity):
+    """Create the section describing a global numbering.
+
+    :arg mesh: The extruded mesh.
+    :arg nodes_per_entity: Number of nodes on, and on top of, each
+        type of topological entity on the base mesh for a single cell
+        layer.  Multiplying up by the number of layers happens in this
+        function.
+
+    :returns: A PETSc Section providing the number of dofs, and offset
+        of each dof, on each mesh point.
+    """
+    cdef:
+        PETSc.DM dm
+        PETSc.Section section
+        PETSc.IS renumbering
+        PetscInt i, p, layers, pStart, pEnd
+        PetscInt dimension, ndof
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] nodes
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents
+        bint variable
+
+    try:
+        variable = not mesh.cell_set.constant_layers
+    except AttributeError:
+        variable = False
+
+    if variable:
+        layer_extents = mesh.layer_extents
+
+    dm = mesh._plex
+    renumbering = mesh._plex_renumbering
+    section = PETSc.Section().create(comm=mesh.comm)
+    pStart, pEnd = dm.getChart()
+    section.setChart(pStart, pEnd)
+    CHKERR(PetscSectionSetPermutation(section.sec, renumbering.iset))
+
+    nodes = numpy.asarray(nodes_per_entity, dtype=IntType)
+    dimension = dm.getDimension()
+
+    for i in range(dimension + 1):
+        pStart, pEnd = dm.getDepthStratum(i)
+        for p in range(pStart, pEnd):
+            if variable:
+                layers = layer_extents[p, 1] - layer_extents[p, 0]
+                ndof = layers*nodes[i, 0] + (layers - 1)*nodes[i, 1]
+            else:
+                ndof = nodes[i]
```

Comment:
This branching is the only thing different to the `dmplex.create_section` implementation, ripe for merging.